gguf-py/gguf/gguf_manager.py

```diff
@@ -0,0 +1,543 @@
+
+        elif split_arguments.split_style == SplitStyle.SIZE:
+            shards = deque()
+
+            # we have to determine the shards first to determine how many shards there will be in total - two passes
+            for i, shard in enumerate(model):
+                if i == 0:
+                    shards.append([shard])
+                    continue
+                if SplitStrategy.get_tensor_size(shard[1]) + sum(SplitStrategy.get_tensor_size(t[1]) for t in shards[-1]) > split_arguments.split_max_size:
+                    shards.append([shard])
+                else:
+                    shards[-1].append(shard)
+
+            total_shards = len(shards) + split_arguments.small_first_shard
+            shard_offset = 1
+
+            if split_arguments.small_first_shard:
+                outname = fname_out.with_name(SHARD_NAME_FORMAT.format(fname_out.stem, shard_offset, total_shards))
+                self.append((outname, None, GGUFWriter(outname, arch, use_temp_file=use_temp_file, endianess=endianess)))
+                shard_offset += 1
+
+            for i, shard in enumerate(shards):
+                outname = fname_out.with_name(SHARD_NAME_FORMAT.format(fname_out.stem, i + shard_offset, total_shards))
+                self.append((outname, deque(shard), GGUFWriter(outname, arch, use_temp_file=use_temp_file, endianess=endianess)))
+
+    @staticmethod
+    def get_tensor_size(tensor) -> int:
+        # we don't have the LazyTensor class here from convert.py but we can try
+        try:
+            return tensor.data_type.elements_to_bytes(np.prod(tensor.shape))
+        except AttributeError: # numpy ndarray[Any, Any]
+            return tensor.nbytes
+        except: # this should never happen
+            raise ValueError(f"Invalid tensor type: {type(tensor)}")
+    
+    @staticmethod
+    def split_str_to_n_bytes(split_str: str) -> int:
+        if split_str.endswith("K"):
+            n = int(split_str[:-1]) * 1024
+        elif split_str.endswith("M"):
+            n = int(split_str[:-1]) * 1024 * 1024
+        elif split_str.endswith("G"):
+            n = int(split_str[:-1]) * 1024 * 1024 * 1024
+        elif split_str.isnumeric():
+            n = int(split_str)
+        else:
+            raise ValueError(f"Invalid split size: {split_str}, must be a number, optionally followed by K, M, or G")
+
+        if n <= 0:
+            raise ValueError(f"Invalid split size: {split_str}, must be positive")
+
+        return n
+
+    @staticmethod
+    def format_n_bytes_to_str(num: int) -> str:
+        num = float(num)
+        for unit in ("", "K", "M", "G"):
+            if abs(num) < 1024.0:
+                return f"{num:3.1f}{unit}"
+            num /= 1024.0
+        return f"{num:.1f}T - over 1TB, --split recommended"
+
+
+# ideally this has most of the same signatures as GGUFWriter so it's nearly a drop-in replacement
```

Comment:
It's likely possible to duplicate even less of the methods by doing something similar to the Lazy tensors from `gguf-py/gguf/lazy.py`. Something with `__getattr__` to intercept missing attributes and redirect them.

But it might make `kv_data` management more complicated, though.

Reply:
Not familiar with how those work (other than they gave me a memory leak, haha) but I'll take a look. Another thought on reducing boilerplate - could we have `GGUFManager` subclass `GGUFWriter` somehow?